### PR TITLE
✨(apps) add the grafana application

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Grafana is a popular visualization and analytics software used to mine data.

## Proposal

A versatile tray has been implemented to ease grafana provisioning in various conditions.

### CODE REVIEW

File: apps/grafana/vars/all/main.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+grafana_database_secret_name: "grafana-database-{{ grafana_vault_checksum | default('undefined_grafana_vault_checksum') }}"
+
+# -- grafana
+grafana_image_name: "grafana/grafana"
+grafana_image_tag: "8.0.3"
+grafana_container_uid: 472
+grafana_container_gid: 0
+# The image pull secret name should match the name of your secret created to
+# login to your private docker registry
+grafana_image_pull_secret_name: ""
```
Comment (quitterie-lcs) [inline apps/grafana/vars/all/main.yml:22]: Not sure to understand, where is this match checked?

File: apps/grafana/vars/vault/main.yml.j2 [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+DB_NAME: {{ postgresql_credentials.name }}
+DB_USER: {{ postgresql_credentials.user }}
+DB_PASSWORD: {{ postgresql_credentials.password }}
+
+# Grafana
+GF_DATABASE_NAME: {{ postgresql_credentials.name }}
+GF_DATABASE_USER: {{ postgresql_credentials.user }}
+GF_DATABASE_PASSWORD: {{ postgresql_credentials.password }}
+# GF_SECURITY_ADMIN_PASSWORD:
```
Comment (quitterie-lcs) [inline apps/grafana/vars/vault/main.yml.j2:15]: As this variable is not used, can it be removed ?
```suggestion
```

Comment (jmaupetit) [reply apps/grafana/vars/vault/main.yml.j2:15]: This vault is an example of what could/should be added to secrets!

File: apps/grafana/vars/all/main.yml (hunk 1 above)
Comment (jmaupetit) [reply apps/grafana/vars/all/main.yml:22]: It is used here in the deployment: https://github.com/openfun/arnold-apps/pull/8/files#diff-deb9262786842ff19dea29f66db2f2418415d9b446008a24967a5bdd4d8e89c2R42

File: apps/grafana/templates/services/app/deploy.yml.j2 [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            - weight: 100
+              podAffinityTerm:
+                labelSelector:
+                  matchExpressions:
+                    - key: deployment
+                      operator: In
+                      values:
+                        - "grafana-app-{{ deployment_stamp }}"
+                topologyKey: kubernetes.io/hostname
+{% set image_pull_secret_name = grafana_image_pull_secret_name | default(none) or default_image_pull_secret_name %}
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply apps/grafana/templates/services/app/deploy.yml.j2:42]: if `grafana_image_pull_secret_name == ""` then `grafana_image_pull_secret_name | default(none)` equal `None` and thus `image_pull_secret_name` equal `default_image_pull_secret_name`. I've just tested this point.

File: apps/grafana/templates/services/app/deploy.yml.j2 [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            - mountPath: /var/lib/grafana
+              name: grafana-resources
+            - name: grafana-provisioning-datasources
+              mountPath: /etc/grafana/provisioning/datasources/datasources.yml
+              subPath: datasources.yml
+            - name: grafana-provisioning-dashboards
+              mountPath: /etc/grafana/provisioning/dashboards/dashboards.yml
+              subPath: dashboards.yml
+      securityContext:
+        runAsUser: {{ grafana_container_uid }}
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply apps/grafana/templates/services/app/deploy.yml.j2:97]: From grafana's documentation, it requires a specific/fixed UID. IMO it should be different from the generic container UID used in Arnold, as its app/service-specific.

File: apps/grafana/vars/all/main.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
+# Application default configuration
+
+# -- route
+grafana_host: "grafana.{{ namespace_name }}.{{ domain_name }}"
+
+# -- databases (defaults to postgresql)
+grafana_postgresql_version: "13.3"
+grafana_database_engine: "postgres"
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply apps/grafana/vars/all/main.yml:8]: Good point!

File: apps/grafana/vars/all/main.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+grafana_image_tag: "8.0.3"
+grafana_container_uid: 472
+grafana_container_gid: 0
+# The image pull secret name should match the name of your secret created to
+# login to your private docker registry
+grafana_image_pull_secret_name: ""
+grafana_app_port: 3000
+grafana_app_replicas: 1
+grafana_secret_name: "grafana-{{ grafana_vault_checksum | default('undefined_grafana_vault_checksum') }}"
+grafana_activate_http_basic_auth: false
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply apps/grafana/vars/all/main.yml:26]: It's not used. Should be removed. Thanks!

**Comment by jmaupetit** [comment]
@quitterie-lcs & @lunika I've addressed all your comments. Ready for a second round? :pray:

**Review by lunika — APPROVED** [review]

**Pull request merged**